src/test/java/uk/gov/justice/hmpps/prison/api/resource/impl/BookingResourceIntTest.kt

```diff
@@ -0,0 +1,1132 @@
+    assertThat(response.body).isEqualTo("[]")
+  }
+
+  @Test
+  fun offenceHistory() {
+    val response = testRestTemplate.exchange(
+      "/api/bookings/offenderNo/{offenderNo}/offenceHistory",
+      GET,
+      createHttpEntity(VIEW_PRISONER_DATA, null),
+      String::class.java,
+      "A1234AG",
+    )
+    assertThatJsonFileAndStatus(response, 200, "offender_main_offences.json")
+  }
+
+  @Test
+  fun offenceHistoryIncludeOffenderWithoutConviction() {
+    val response = testRestTemplate.exchange(
+      "/api/bookings/offenderNo/{offenderNo}/offenceHistory?convictionsOnly=false",
+      GET,
+      createHttpEntity(VIEW_PRISONER_DATA, null),
+      String::class.java,
+      "A1234AB",
+    )
+    assertThatJsonFileAndStatus(response, 200, "offender_offence_history_A12234AB_include_non_convictions.json")
+  }
+
+  @Test
+  fun offenceHistoryForBookings() {
+    val token = validToken(listOf("ROLE_VIEW_PRISONER_DATA"))
+    val httpEntity = createHttpEntity(token, java.util.List.of(-59, -7, -3))
+    val response = testRestTemplate.exchange(
+      "/api/bookings/offence-history",
+      POST,
+      httpEntity,
+      object : ParameterizedTypeReference<String?>() {},
+    )
+    assertThatJsonFileAndStatus(response, 200, "offence-history-by-bookingids.json")
+  }
+
+  @Test
+  fun secondaryLanguages() {
+    val response = testRestTemplate.exchange(
+      "/api/bookings/{bookingId}/secondary-languages",
+      GET,
+      createHttpEntity(NORMAL_USER, null),
+      String::class.java,
+      -3L,
+    )
+    assertThatJsonFileAndStatus(response, 200, "secondary_languages.json")
+  }
+
+  @Test
+  fun nextVisit() {
+    val response = testRestTemplate.exchange(
+      "/api/bookings/{bookingId}/visits/next",
+      GET,
+      createHttpEntity(NORMAL_USER, null),
+      String::class.java,
+      -3L,
+    )
+    assertThatJsonFileAndStatus(response, 200, "next-visit.json")
+  }
+
+  @Test
+  fun nextVisit_withVisitors() {
+    val response = testRestTemplate.exchange(
+      "/api/bookings/{bookingId}/visits/next?withVisitors=true",
+      GET,
+      createHttpEntity(NORMAL_USER, null),
+      String::class.java,
+      -3L,
+    )
+    assertThatJsonFileAndStatus(response, 200, "next-visit-with-visitors.json")
+  }
+
+  @Test
+  fun nextVisit_withVisitors_whenNotPresent() {
+    val response = testRestTemplate.exchange(
+      "/api/bookings/{bookingId}/visits/next?withVisitors=true",
+      GET,
+      createHttpEntity(NORMAL_USER, null),
+      String::class.java,
+      -1,
+    )
+    assertThatStatus(response, 200)
+  }
+
+  @Test
+  fun nextVisit_whenNotPresent() {
+    val response = testRestTemplate.exchange(
+      "/api/bookings/{bookingId}/visits/next",
+      GET,
+      createHttpEntity(NORMAL_USER, null),
+      String::class.java,
+      -1,
+    )
+    assertThatStatus(response, 200)
+  }
+
+  @Nested
+  @DisplayName("GET /api/bookings/{bookingId}/visits-with-visitors")
+  inner class VisitsWithVisitors {
```

Comment:
Added authorisation tests for updated endpoint (which now also has an alternative role)